system/inc/active_object.h

```diff
@@ -0,0 +1,335 @@
+     * @param obj
+     */
+    static void run_active_object(ActiveObjectBase* obj);
+
+    void start_thread();
+
+public:
+
+    ActiveObjectBase(const ActiveObjectConfiguration& config) : configuration(config), started(false) {}
+
+    bool isCurrentThread() {
+        return _thread_id == std::this_thread::get_id();
+    }
+
+    bool isStarted() {
+        return started;
+    }
+
+    template<typename arg, typename r> inline void invoke(r (*fn)(arg* a), arg* value, size_t size) {
+        //invoke_impl((void*)fn, value, size);
+    }
+
+    template<typename arg, typename r> inline void invoke(r (*fn)(arg* a), arg* value) {
+        //invoke_impl((void*)fn, value);
+    }
+
+    void invoke(void (*fn)()) {
+        //invoke_impl((void*)fn, NULL, 0);
+    }
+
+    template<typename R> std::shared_ptr<Promise<R>> invoke_future(std::function<R(void)> work)
+    {
+        auto promise = std::make_shared<Promise<R>>(work);            // shared pointer for reference returned here
+        Item message = std::static_pointer_cast<Message>(promise);    // another reference to put on the queue
+        put(message);
+        return promise;
+    }
+
+};
+
+
+template <size_t queue_size=50>
+class ActiveObjectChannel : public ActiveObjectBase
+{
+    cpp::channel<Item, queue_size> _channel;
+
+protected:
+
+    virtual bool take(Item& item) override
+    {
+        return cpp::select().recv_only(_channel, item).try_once();
+    }
+
+    virtual void put(const Item& item) override
+    {
+        _channel.send(item);
+    }
+
+
+public:
+
+    ActiveObjectChannel(ActiveObjectConfiguration& config) : ActiveObjectBase(config) {}
+
+    /**
+     * Start the asynchronous processing for this active object.
+     */
+    void start()
+    {
+        _channel = cpp::channel<Item*, queue_size>();
+        start_thread();
+    }
+
+};
+
+class ActiveObjectQueue : public ActiveObjectBase
+{
+    os_queue_t  queue;
+
+protected:
+
+    virtual bool take(Item& result)
+    {
+        return os_queue_take(queue, &result, configuration.take_wait)==0;
+    }
+
+    virtual void put(Item& item)
+    {
+        // since the queue does a binary copy rather than using the assignment operator
+        // we don't get the required shared pointer reference increment
+        // force it by doing a placement new allocation and invoke the copy ctor
+        uint8_t buf[sizeof(Item)];
+        Item* pItem = new (buf) Item(std::move(item));     // placement new, and copy construct from item
+        while (os_queue_put(queue, pItem, CONCURRENT_WAIT_FOREVER)) {}
+        // item is reset when removed from the queue.
+    }
+
+    void createQueue(int size=50)
+    {
+        os_queue_create(&queue, sizeof(Item*), size);
+    }
+
+public:
+
+    ActiveObjectQueue(const ActiveObjectConfiguration& config) : ActiveObjectBase(config), queue(NULL) {}
+
+    void start()
+    {
+        createQueue();
+    }
+};
+
+
+class ActiveObjectCurrentThreadQueue : public ActiveObjectQueue
```

Comment:
How about ActiveObjectQueueInCurrentThread and ActiveObjectQueueInSeparateThread?